### ISSUE TITLE
ValueError: A non-empty list for args is required

### ISSUE THREAD

**Issue by stevie004** [h-hirokawa/atom-autocomplete-ansible#44]
OS: Fedora 31
Atom: 1.41.0 
ansible --version
ansible 2.9.1
  config file = None
  configured module search path = ['/home/stefan41/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/stefan41/.local/lib/python3.7/site-packages/ansible
  executable location = /home/stefan41/.local/bin/ansible
  python version = 3.7.5 (default, Oct 17 2019, 12:16:48) [GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]

Following error occurs when starting Atom:

```
autocomplete-ansible traceback output:

Command failed: "/usr/bin/python" "/home/stefan41/.atom/packages/autocomplete-ansible/libs/parse_ansible.py"
Traceback (most recent call last):
  File "/home/stefan41/.atom/packages/autocomplete-ansible/libs/parse_ansible.py", line 35, in <module>
    doc_cli = DocCLI([])
  File "/home/stefan41/.local/lib/python3.7/site-packages/ansible/cli/doc.py", line 58, in __init__
    super(DocCLI, self).__init__(args)
  File "/home/stefan41/.local/lib/python3.7/site-packages/ansible/cli/__init__.py", line 67, in __init__
    raise ValueError('A non-empty list for args is required')
ValueError: A non-empty list for args is required
```

**Comment by dnorthup-ums** [comment]
Ditto for the following:

autocomplete-ansible version 0.1.24

```bash
✔ ~/local_git_repos/.../Ansible_Roles/u_ [mole_c8_init|✔] 
14:21 $ cat /etc/fedora-release 
Fedora release 30 (Thirty)
✔ ~/local_git_repos/.../Ansible_Roles/u_ [mole_c8_init|✔] 
14:21 $ ansible --version
ansible 2.9.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/dnorthup/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.5 (default, Oct 17 2019, 12:09:47) [GCC 9.2.1 20190827 (Red Hat 9.2.1-1)]
✔ ~/local_git_repos/.../Ansible_Roles/u_ [mole_c8_init|✔] 
14:21 $ rpm -qi atom
Name        : atom
Version     : 1.40.0
Release     : 0.1
Architecture: x86_64
Install Date: Tue 13 Aug 2019 10:24:13 AM EDT
Group       : Unspecified
Size        : 570783704
License     : MIT
Signature   : (none)
Source RPM  : atom-1.40.0-0.1.src.rpm
Build Date  : Fri 09 Aug 2019 11:36:31 AM EDT
Build Host  : b01bbeaf3a88
Relocations : /usr 
URL         : https://atom.io/
Summary     : A hackable text editor for the 21st Century.
Description :
A hackable text editor for the 21st Century.
```

**Comment by jpartain89** [comment]
Mine would make three... I really loved this add on and until this is fixed, I'm at a loss for anything that comes close!

This is the error message that I get:

```bash
autocomplete-ansible traceback output:

Command failed: "/usr/local/opt/python3/bin/python3" "/Users/jpartain89/.atom/packages/autocomplete-ansible/libs/parse_ansible.py"
Traceback (most recent call last):
  File "/Users/jpartain89/.atom/packages/autocomplete-ansible/libs/parse_ansible.py", line 35, in <module>
    doc_cli = DocCLI([])
  File "/usr/local/lib/python3.7/site-packages/ansible/cli/doc.py", line 58, in __init__
    super(DocCLI, self).__init__(args)
  File "/usr/local/lib/python3.7/site-packages/ansible/cli/__init__.py", line 67, in __init__
    raise ValueError('A non-empty list for args is required')
ValueError: A non-empty list for args is required
```

```bash
$ head -1 /usr/local/bin/ansible
#!/usr/local/opt/python/bin/python3.7
$ ansible --version
ansible 2.9.1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/Users/jpartain89/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.7.5 (default, Nov  1 2019, 02:16:23) [Clang 11.0.0 (clang-1100.0.33.8)]
```

**Comment by ebenezar-mccoy** [comment]
is this project dead ?

**Comment by dnorthup-ums** [comment]
@ebenezar-mccoy I'm not sure, but the integration tests haven't been updated since September of 2018... If I fix that I get errors like the following:

```bash
...
_______________________________________ ERROR collecting tests/test_parse_ansible.py ________________________________________
tests/test_parse_ansible.py:3: in <module>
    from libs import parse_ansible
libs/parse_ansible.py:35: in <module>
    doc_cli = DocCLI([])
.tox/py38-ansible29/lib/python3.8/site-packages/ansible/cli/doc.py:58: in __init__
    super(DocCLI, self).__init__(args)
.tox/py38-ansible29/lib/python3.8/site-packages/ansible/cli/__init__.py:67: in __init__
    raise ValueError('A non-empty list for args is required')
E   ValueError: A non-empty list for args is required
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================================= 1 error in 0.80s ===========================================
ERROR: InvocationError for command '/home/.../local_git_repos/atom-autocomplete-ansible/.tox/py38-ansible29/bin/pytest' (exited with code 2)
__________________________________________ summary ____________________________________________________
  py27-ansible27: commands succeeded
  py27-ansible28: commands succeeded
ERROR:   py27-ansible29: commands failed
  py35-ansible27: commands succeeded
  py35-ansible28: commands succeeded
ERROR:   py35-ansible29: commands failed
  py36-ansible27: commands succeeded
  py36-ansible28: commands succeeded
ERROR:   py36-ansible29: commands failed
  py37-ansible27: commands succeeded
  py37-ansible28: commands succeeded
ERROR:   py37-ansible29: commands failed
  py38-ansible27: commands succeeded
  py38-ansible28: commands succeeded
ERROR:   py38-ansible29: commands failed
```

@ALL
This seems to be due to a change in the \_\_init\_\_.py superclass of all of the CLI programs in Ansible. There's likely a work-around (haven't found it yet), but I'm not expecting them to do anything about this from the Ansible side.

**Comment by ebenezar-mccoy** [comment]
@ALL temporary workaround:

`❯ git show 8dd83c4
8dd83c4 74 seconds ago Mariusz Makowski (HEAD -> feature-ansible-2.9) ansible 2.9 fix
diff --git a/libs/parse_ansible.py b/libs/parse_ansible.py
index cdc3862..025b36e 100755
--- a/libs/parse_ansible.py
+++ b/libs/parse_ansible.py
@@ -32,7 +32,7 @@ except ImportError:
     USE_FRAGMENT_LOADER = False

 __main__.display = Display()
-doc_cli = DocCLI([])
+doc_cli = DocCLI(['ansible atom'])


 def get_module_list():
diff --git a/libs/provider.js b/libs/provider.js
index f59c6c5..199e56d 100644
--- a/libs/provider.js
+++ b/libs/provider.js
@@ -37,7 +37,7 @@ export default class Provider {
     const interpreter = interpretersLookup.getInterpreter();
     const script = pathJoin(__dirname, 'parse_ansible.py');
     exec(`"${interpreter}" "${script}"`,
-      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 10*1024*1024},
+      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 20*1024*1024},
       function (err, stdout) {
         self._loading = false;
         if (err) {
`

**Comment by ebenezar-mccoy** [comment]
@ALL, fix:

```
diff --git a/libs/parse_ansible.py b/libs/parse_ansible.py
index cdc3862..025b36e 100755
--- a/libs/parse_ansible.py
+++ b/libs/parse_ansible.py
@@ -32,7 +32,7 @@ except ImportError:
     USE_FRAGMENT_LOADER = False

 __main__.display = Display()
-doc_cli = DocCLI([])
+doc_cli = DocCLI(['ansible atom'])


 def get_module_list():
diff --git a/libs/provider.js b/libs/provider.js
index f59c6c5..199e56d 100644
--- a/libs/provider.js
+++ b/libs/provider.js
@@ -37,7 +37,7 @@ export default class Provider {
     const interpreter = interpretersLookup.getInterpreter();
     const script = pathJoin(__dirname, 'parse_ansible.py');
     exec(`"${interpreter}" "${script}"`,
-      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 10*1024*1024},
+      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 20*1024*1024},
       function (err, stdout) {
         self._loading = false;
         if (err) {

```

#45

**Comment by dnorthup-ums** [comment]
This is basically what I tried first (not exactly identical, but close enough). While it passes the automated tests it fails with:
"Uncaught NodeError: stdout maxBuffer length exceeded
    at Socket.onChildStdout (child_process.js:357:14)
    at Socket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:279:12)
    at readableAddChunk (_stream_readable.js:260:13)
    at Socket.Readable.push (_stream_readable.js:219:10)
    at Pipe.onread (net.js:636:20)"

at "provider.js? [sm]:67" (which is the line where uncaught errors are thrown).

That's why I didn't tell anybody to use it.

> @ALL, fix:
> 
> ```
> diff --git a/libs/parse_ansible.py b/libs/parse_ansible.py
> index cdc3862..025b36e 100755
> --- a/libs/parse_ansible.py
> +++ b/libs/parse_ansible.py
> @@ -32,7 +32,7 @@ except ImportError:
>      USE_FRAGMENT_LOADER = False
> 
>  __main__.display = Display()
> -doc_cli = DocCLI([])
> +doc_cli = DocCLI(['ansible atom'])
> 
> 
>  def get_module_list():
> diff --git a/libs/provider.js b/libs/provider.js
> index f59c6c5..199e56d 100644
> --- a/libs/provider.js
> +++ b/libs/provider.js
> @@ -37,7 +37,7 @@ export default class Provider {
>      const interpreter = interpretersLookup.getInterpreter();
>      const script = pathJoin(__dirname, 'parse_ansible.py');
>      exec(`"${interpreter}" "${script}"`,
> -      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 10*1024*1024},
> +      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 20*1024*1024},
>        function (err, stdout) {
>          self._loading = false;
>          if (err) {
> ```
> 
> #45

**Comment by ebenezar-mccoy** [comment]
@dnorthup-ums fix above passes the tests and works in atom for me.

with ansible 2.9 you have to increase maxBuffer in js executor - because, ansible documentation is returning more objects now

```
-      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 10*1024*1024},
+      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 20*1024*1024},
```

**Comment by dnorthup-ums** [comment]
@ebenezar-mccoy I did in fact try that. Apparently I had to stop and restart atom twice (?!!!!!) before it would let me test your fix.

I'm fairly sure that it isn't supposed to be returning 20MiB worth of text as well, but that's not your fault.

Objection dropped.

**Comment by Pax-nl** [comment]
Hi,

I updated

```
-      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 10*1024*1024},
+      {env: {ANSIBLE_DEPRECATION_WARNINGS: 'False'}, maxBuffer: 20*1024*1024},
```

Which solved the issue on my system:

Mac Os Catalina, Ansible 2.9.2, Atom 1.42.0

I also updated the doc_cli but this was not needed for the fix..

```
-doc_cli = DocCLI([])
+doc_cli = DocCLI(['ansible atom'])
```

@ebenezar-mccoy hoping your PR will be merged soon.. thank you for looking into this..

**Comment by dnorthup-ums** [comment]
@h-hirokawa Is there a chance you can merge the pull requests attached to this issue or give somebody else the permission to do so (as this is the repository that Atom is using as a source for this plugin)?

**Comment by Pax-nl** [comment]
> @h-hirokawa Is there a chance you can merge the pull requests attached to this issue or give somebody else the permission to do so (as this is the repository that Atom is using as a source for this plugin)?

+1

**Comment by h-hirokawa** [comment]
Really sorry for the delay on this.
I have merged #45 and #47, so the master branch should work fine with Ansible  2.9.
However, I haven't used Atom anymore, so I haven't tried it actually.

Is there anyone who wants to take over the maintainer of this project?

**Comment by VeselaHouba** [comment]
@h-hirokawa Update is not yet visible in atom packaging system. I'm not sure what needs to be done before atom.io picks up changes and propagates them as update. Maybe tag it as release?

**Comment by dnorthup-ums** [comment]
> Really sorry for the delay on this.
> I have merged #45 and #47, so the master branch should work fine with Ansible 2.9.
> However, I haven't used Atom anymore, so I haven't tried it actually.
> 
> Is there anyone who wants to take over the maintainer of this project?

@h-hirokawa 
This seems to be fairly low maintenance so, unless @ebenezar-mccoy wants to do it more than I do, I'm game (and I've done the whole open source benevolent dictator thing before). I presume we'll need to let the Atom editor people know?

**Comment by dv8472** [comment]
A simple workaround (based on [solution above](https://github.com/h-hirokawa/atom-autocomplete-ansible/issues/44#issuecomment-563425387)) until oficial fix is released.
For Arch Linux (I don't know where other distros store their atom files, to search for it in your $HOME use `find ~ -name parse_ansible.py -o -name provider.js`, then adjust the path accordingly, in the commands below):
```
sed "s/\(doc_cli = DocCLI(\[\)/\1'ansible atom'/" -i ~/.atom/packages/autocomplete-ansible/libs/parse_ansible.py
sed "/ANSIBLE_DEPRECATION_WARNINGS/s/\(maxBuffer: \)10/\120/" -i ~/.atom/packages/autocomplete-ansible/libs/provider.js
```

**Comment by dnorthup-ums** [comment]
@h-hirokawa has given me permission to collaborate on this project. I'm seeing about getting a release tagged and on the master branch. Hopefully we can then get this fix out to all Atom users. Given that, I'm closing this issue.
